<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate scores for each amino acid position in a protein sequence — calculate_aa_scores • protti</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate scores for each amino acid position in a protein sequence — calculate_aa_scores"><meta property="og:description" content="
Calculate a score for each amino acid position in a protein sequence based on the product of the
-log10(adjusted p-value) and the absolute log2(fold change) per peptide covering this amino acid. In detail, all the
peptides are aligned along the sequence of the corresponding protein, and the average score per
amino acid position is computed. In a limited proteolysis coupled to mass spectrometry (LiP-MS)
experiment, the score allows to prioritize and narrow down structurally affected regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">protti</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_analysis_dose_response_workflow.html">Dose-Response Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/data_analysis_single_dose_treatment_workflow.html">Single Dose Treatment Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/input_preparation_workflow.html">Input Preparation Workflow</a>
    </li>
    <li>
      <a href="../articles/protein_structure_workflow.html">Protein Structure Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/quality_control_workflow.html">Qualtiy Control (QC) Workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jpquast/protti/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate scores for each amino acid position in a protein sequence</h1>
    <small class="dont-index">Source: <a href="https://github.com/jpquast/protti/blob/HEAD/R/calculate_aa_scores.R" class="external-link"><code>R/calculate_aa_scores.R</code></a></small>
    <div class="hidden name"><code>calculate_aa_scores.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a>
Calculate a score for each amino acid position in a protein sequence based on the product of the
-log10(adjusted p-value) and the absolute log2(fold change) per peptide covering this amino acid. In detail, all the
peptides are aligned along the sequence of the corresponding protein, and the average score per
amino acid position is computed. In a limited proteolysis coupled to mass spectrometry (LiP-MS)
experiment, the score allows to prioritize and narrow down structurally affected regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculate_aa_scores</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">protein</span>,</span>
<span>  diff <span class="op">=</span> <span class="va">diff</span>,</span>
<span>  adj_pval <span class="op">=</span> <span class="va">adj_pval</span>,</span>
<span>  <span class="va">start_position</span>,</span>
<span>  <span class="va">end_position</span>,</span>
<span>  retain_columns <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>a data frame containing at least the input columns.</p></dd>


<dt>protein</dt>
<dd><p>a character column in the data frame containing the protein identifier or name.</p></dd>


<dt>diff</dt>
<dd><p>a numeric column in the <code>data</code> data frame containing the log2 fold change.</p></dd>


<dt>adj_pval</dt>
<dd><p>a numeric column in the <code>data</code> data frame containing the adjusted p-value.</p></dd>


<dt>start_position</dt>
<dd><p>a numeric column <code>data</code> in the data frame containing the start position
of a peptide or precursor.</p></dd>


<dt>end_position</dt>
<dd><p>a numeric column in the data frame containing the end position of a peptide or
precursor.</p></dd>


<dt>retain_columns</dt>
<dd><p>a vector indicating if certain columns should be retained from the input
data frame. Default is not retaining additional columns <code>retain_columns = NULL</code>. Specific
columns can be retained by providing their names (not in quotations marks, just like other
column names, but in a vector).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame that contains the aggregated scores per amino acid position, enabling to
draw fingerprints for each individual protein.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Patrick Stalder</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  pg_protein_accessions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"protein_1"</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  diff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="op">-</span><span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  adj_pval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.2</span>, <span class="fl">0.05</span>, <span class="fl">0.002</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.7</span>, <span class="fl">0.001</span>, <span class="fl">0.02</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">15</span>, <span class="fl">25</span>, <span class="fl">28</span>, <span class="fl">30</span>, <span class="fl">41</span>, <span class="fl">51</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">16</span>, <span class="fl">23</span>, <span class="fl">35</span>, <span class="fl">35</span>, <span class="fl">35</span>, <span class="fl">48</span>, <span class="fl">55</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">calculate_aa_scores</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data</span>,</span></span>
<span class="r-in"><span>  protein <span class="op">=</span> <span class="va">pg_protein_accessions</span>,</span></span>
<span class="r-in"><span>  diff <span class="op">=</span> <span class="va">diff</span>,</span></span>
<span class="r-in"><span>  adj_pval <span class="op">=</span> <span class="va">adj_pval</span>,</span></span>
<span class="r-in"><span>  start_position <span class="op">=</span> <span class="va">start</span>,</span></span>
<span class="r-in"><span>  end_position <span class="op">=</span> <span class="va">end</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 47 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Groups:   pg_protein_accessions, residue [47]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    pg_protein_accessions residue amino_acid_score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> protein_1                   1            6    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> protein_1                   2            6    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> protein_1                   3            6    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> protein_1                   4            6    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> protein_1                   5            4.23 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> protein_1                   6            4.23 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> protein_1                   7            3.35 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> protein_1                   8            3.35 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> protein_1                   9            0.699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> protein_1                  10            1.65 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 37 more rows</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan-Philipp Quast, Dina Schuster, ETH Zurich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

