<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform gene ontology enrichment analysis — calculate_go_enrichment • protti</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform gene ontology enrichment analysis — calculate_go_enrichment"><meta property="og:description" content="Analyses enrichment of gene ontology terms associated with proteins in the fraction of
significant proteins compared to all detected proteins. A two-sided Fisher's exact test is
performed to test significance of enrichment or depletion. GO annotations can be provided to
this function either through UniProt go_annotations_uniprot, through a table obtained
with fetch_go in the go_data argument or GO annotations are fetched automatically
by the function by providing ontology_type and organism_id."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">protti</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_analysis_dose_response_workflow.html">Dose-Response Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/data_analysis_single_dose_treatment_workflow.html">Single Dose Treatment Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/input_preparation_workflow.html">Input Preparation Workflow</a>
    </li>
    <li>
      <a href="../articles/protein_structure_workflow.html">Protein Structure Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/quality_control_workflow.html">Qualtiy Control (QC) Workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jpquast/protti/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform gene ontology enrichment analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/jpquast/protti/blob/HEAD/R/calculate_go_enrichment.R" class="external-link"><code>R/calculate_go_enrichment.R</code></a></small>
    <div class="hidden name"><code>calculate_go_enrichment.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Analyses enrichment of gene ontology terms associated with proteins in the fraction of
significant proteins compared to all detected proteins. A two-sided Fisher's exact test is
performed to test significance of enrichment or depletion. GO annotations can be provided to
this function either through UniProt <code>go_annotations_uniprot</code>, through a table obtained
with <code>fetch_go</code> in the <code>go_data</code> argument or GO annotations are fetched automatically
by the function by providing <code>ontology_type</code> and <code>organism_id</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculate_go_enrichment</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">protein_id</span>,</span>
<span>  <span class="va">is_significant</span>,</span>
<span>  go_annotations_uniprot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">ontology_type</span>,</span>
<span>  organism_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  go_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_cutoff <span class="op">=</span> <span class="st">"adj_pval top10"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>a data frame that contains at least the input variables.</p></dd>


<dt>protein_id</dt>
<dd><p>a character column in the <code>data</code> data frame that contains the protein
accession numbers.</p></dd>


<dt>is_significant</dt>
<dd><p>a logical column in the <code>data</code> data frame that indicates if the
corresponding protein has a significantly changing peptide. The input data frame may contain
peptide level information with significance information. The function is able to extract
protein level information from this.</p></dd>


<dt>go_annotations_uniprot</dt>
<dd><p>recommended, a character column in the <code>data</code> data frame
that contains gene ontology annotations obtained from UniProt using <code>fetch_uniprot</code>.
These annotations are already separated into the desired ontology type so the argument
<code>ontology_type</code> is not required.</p></dd>


<dt>ontology_type</dt>
<dd><p>optional, character value specifying the type of ontology that should
be used. Possible values are molecular function (MF), biological process (BP), cellular component
(CC). This argument is not required if GO annotations are provided from UniProt in
<code>go_annotations_uniprot</code>. It is required if annotations are provided through <code>go_data</code> or
automatically fetched.</p></dd>


<dt>organism_id</dt>
<dd><p>optional, character value specifying an NCBI taxonomy identifier of an
organism (TaxId). Possible inputs include only: "9606" (Human), "559292" (Yeast) and "83333"
(E. coli). Is only necessary if GO data is not provided either by <code>go_annotations_uniprot</code>
or in <code>go_data</code>.</p></dd>


<dt>go_data</dt>
<dd><p>Optional, a data frame that can be obtained with <code>fetch_go</code>. If you provide
data not obtained with <code>fetch_go</code> make sure column names for protein ID (db_id) and GO ID
(go_id) are the same as for data obtained with <code>fetch_go</code>.</p></dd>


<dt>plot</dt>
<dd><p>a logical argument indicating whether the result should be plotted or returned as a table.</p></dd>


<dt>label</dt>
<dd><p>a logical argument indicating whether labels should be added to the plot.
Default is TRUE.</p></dd>


<dt>plot_cutoff</dt>
<dd><p>a character value indicating if the plot should contain the top 10 most
significant proteins (p-value or adjusted p-value), or if a significance cutoff should be used
to determine the number of GO terms in the plot. This information should be provided with the
type first followed by the threshold separated by a space. Example are
<code>plot_cutoff = "adj_pval top10"</code>, <code>plot_cutoff = "pval 0.05"</code> or
<code>plot_cutoff = "adj_pval 0.01"</code>. The threshold can be chosen freely.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A bar plot displaying negative log10 adjusted p-values for the top 10 enriched or
depleted gene ontology terms. Alternatively, plot cutoffs can be chosen individually with the
<code>plot_cutoff</code> argument. Bars are colored according to the direction of the enrichment. If
<code>plot = FALSE</code>, a data frame is returned. P-values are adjusted with Benjamini-Hochberg.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Load libraries</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘dplyr’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     filter, lag</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect, setdiff, setequal, union</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create example data</span></span></span>
<span class="r-in"><span><span class="co"># Contains artificial de-enrichment for ribosomes.</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="fetch_uniprot_proteome.html">fetch_uniprot_proteome</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  organism_id <span class="op">=</span> <span class="fl">83333</span>,</span></span>
<span class="r-in"><span>  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"id"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"go(molecular function)"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="fl">1000</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1000</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="va">go_molecular_function</span>,</span></span>
<span class="r-in"><span>      pattern <span class="op">=</span> <span class="st">"ribosome"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    <span class="va">significant</span></span></span>
<span class="r-in"><span>  <span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plot gene ontology enrichment</span></span></span>
<span class="r-in"><span><span class="fu">calculate_go_enrichment</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data</span>,</span></span>
<span class="r-in"><span>  protein_id <span class="op">=</span> <span class="va">id</span>,</span></span>
<span class="r-in"><span>  go_annotations_uniprot <span class="op">=</span> <span class="va">go_molecular_function</span>,</span></span>
<span class="r-in"><span>  is_significant <span class="op">=</span> <span class="va">significant</span>,</span></span>
<span class="r-in"><span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  plot_cutoff <span class="op">=</span> <span class="st">"pval 0.01"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculate_go_enrichment-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate gene ontology enrichment</span></span></span>
<span class="r-in"><span><span class="va">go_enrichment</span> <span class="op">&lt;-</span> <span class="fu">calculate_go_enrichment</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data</span>,</span></span>
<span class="r-in"><span>  protein_id <span class="op">=</span> <span class="va">id</span>,</span></span>
<span class="r-in"><span>  go_annotations_uniprot <span class="op">=</span> <span class="va">go_molecular_function</span>,</span></span>
<span class="r-in"><span>  is_significant <span class="op">=</span> <span class="va">significant</span>,</span></span>
<span class="r-in"><span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_enrichment</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 10 × 10</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    term                 go_id    pval adj_pval n_detected_prot… n_detected_prot…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> <span style="color: #949494;">"</span>diguanylate cyclas… [GO:… 6.34<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span>  1.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>             <span style="text-decoration: underline;">3</span>261               13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> <span style="color: #949494;">"</span>structural constit… [GO:… 5.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span>  5.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>             <span style="text-decoration: underline;">3</span>261               58</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> <span style="color: #949494;">"</span>rRNA binding <span style="color: #949494;">"</span>      [GO:… 2.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  1.45<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261               44</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> <span style="color: #949494;">"</span>N-acetylmuramoyl-L… [GO:… 6.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  3.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261                5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> <span style="color: #949494;">"</span>carboxypeptidase a… [GO:… 8.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  3.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261                9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> <span style="color: #949494;">"</span>ribosome binding <span style="color: #949494;">"</span>  [GO:… 1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  3.63<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261               28</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> <span style="color: #949494;">"</span>cytochrome bo3 ubi… [GO:… 3.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  7.74<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261                6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> <span style="color: #949494;">"</span>transmembrane sign… [GO:… 3.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  7.74<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261                6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> <span style="color: #949494;">"</span>kinase activity <span style="color: #949494;">"</span>   [GO:… 3.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  7.74<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261               56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> <span style="color: #949494;">"</span>tRNA binding <span style="color: #949494;">"</span>      [GO:… 4.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  8.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>             <span style="text-decoration: underline;">3</span>261               47</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 4 more variables: n_significant_proteins &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_significant_proteins_in_process &lt;int&gt;, n_proteins_expected &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   direction &lt;chr&gt;</span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan-Philipp Quast, Dina Schuster, ETH Zurich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

