<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fetch structure information from RCSB — fetch_pdb • protti</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fetch structure information from RCSB — fetch_pdb"><meta property="og:description" content="Fetches structure metadata from RCSB. If you want to retrieve atom data such as positions, use
the function fetch_pdb_structure()."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">protti</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_analysis_dose_response_workflow.html">Dose-Response Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/data_analysis_single_dose_treatment_workflow.html">Single Dose Treatment Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/input_preparation_workflow.html">Input Preparation Workflow</a>
    </li>
    <li>
      <a href="../articles/protein_structure_workflow.html">Protein Structure Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/quality_control_workflow.html">Qualtiy Control (QC) Workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jpquast/protti/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fetch structure information from RCSB</h1>
    <small class="dont-index">Source: <a href="https://github.com/jpquast/protti/blob/HEAD/R/fetch_pdb.R" class="external-link"><code>R/fetch_pdb.R</code></a></small>
    <div class="hidden name"><code>fetch_pdb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fetches structure metadata from RCSB. If you want to retrieve atom data such as positions, use
the function <code><a href="fetch_pdb_structure.html">fetch_pdb_structure()</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fetch_pdb</span><span class="op">(</span><span class="va">pdb_ids</span>, batchsize <span class="op">=</span> <span class="fl">100</span>, show_progress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>pdb_ids</dt>
<dd><p>a character vector of PDB identifiers.</p></dd>


<dt>batchsize</dt>
<dd><p>a numeric value that specifies the number of structures to be processed in a
single query. Default is 100.</p></dd>


<dt>show_progress</dt>
<dd><p>a logical value that indicates if a progress bar will be shown. Default is
TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame that contains structure metadata for the PDB IDs provided. The data frame
contains some columns that might not be self explanatory.</p><ul><li><p>auth_asym_id: Chain identifier provided by the author of the structure in order to
match the identification used in the publication that describes the structure.</p></li>
<li><p>label_asym_id: Chain identifier following the standardised convention for mmCIF files.</p></li>
<li><p>entity_beg_seq_id, ref_beg_seq_id, length, pdb_sequence: <code>entity_beg_seq_id</code> is a
position in the structure sequence (<code>pdb_sequence</code>) that matches the position given in
<code>ref_beg_seq_id</code>, which is a position within the protein sequence (not included in the
data frame). <code>length</code> identifies the stretch of sequence for which positions match
accordingly between structure and protein sequence. <code>entity_beg_seq_id</code> is a residue ID
based on the standardised convention for mmCIF files.</p></li>
<li><p>auth_seq_id: Residue identifier provided by the author of the structure in order to
match the identification used in the publication that describes the structure. This character
vector has the same length as the <code>pdb_sequence</code> and each position is the identifier for
the matching amino acid position in <code>pdb_sequence</code>. The contained values are not
necessarily numbers and the values do not have to be positive.</p></li>
<li><p>modified_monomer: Is composed of first the composition ID of the modification, followed
by the <code>label_seq_id</code> position. In parenthesis are the parent monomer identifiers as
they appear in the sequence.</p></li>
<li><p>ligand_*: Any column starting with the <code>ligand_*</code> prefix contains information about
the position, identity and donors for ligand binding sites. If there are multiple entities of
ligands they are separated by "|". Specific donor level information is separated by ";".</p></li>
<li><p>secondar_structure: Contains information about helix and sheet secondary structure elements.
Individual regions are separated by ";".</p></li>
<li><p>unmodeled_structure: Contains information about unmodeled or partially modeled regions in
the model. Individual regions are separated by ";".</p></li>
<li><p>auth_seq_id_original: In some cases the sequence positions do not match the number of residues
in the sequence either because positions are missing or duplicated. This always coincides with modified
residues, however does not always occur when there is a modified residue in the sequence. This column
contains the original <code>auth_seq_id</code> information that does not have these positions corrected.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">pdb</span> <span class="op">&lt;-</span> <span class="fu">fetch_pdb</span><span class="op">(</span>pdb_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6HG1"</span>, <span class="st">"1E9I"</span>, <span class="st">"6D3Q"</span>, <span class="st">"4JHW"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2/6] Extract experimental conditions ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.17s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [3/6] Extracting polymer information: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 1/6 UniProt IDs ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.1s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 2/6 UniProt alignment ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.15s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 3/6 Ligand binding sites ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.21s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 4/6 Modified monomers ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.07s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 5/6 Secondary structure ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.08s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -&gt; 6/6 Unmodeled residues ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.08s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [4/6] Correct author sequence positions for some PDB IDs ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> None to correct(0.36s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [5/6] Extract non-polymer information ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.07s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [6/6] Combine information ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE (0.21s)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pdb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 46</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   pdb_ids auth_…¹ label…² refer…³ prote…⁴ refer…⁵ entit…⁶ ref_b…⁷ length pdb_s…⁸</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> 6HG1    A       A       P27708  CAD pr… UniProt     115    <span style="text-decoration: underline;">1</span>571    256 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> 6HG1    A       A       P27708  CAD pr… UniProt     115    <span style="text-decoration: underline;">1</span>571    256 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> 6HG1    A       A       P27708  CAD pr… UniProt       2    <span style="text-decoration: underline;">1</span>460    100 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> 6HG1    A       A       P27708  CAD pr… UniProt       2    <span style="text-decoration: underline;">1</span>460    100 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> 6HG1    A       A       P05020  Dihydr… UniProt     102     107     13 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> 6HG1    A       A       P05020  Dihydr… UniProt     102     107     13 AKLVRL…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 36 more variables: auth_seq_id &lt;chr&gt;, auth_seq_id_original &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   engineered_mutation &lt;chr&gt;, modified_monomer &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ligand_donor_atom_id &lt;chr&gt;, ligand_donor_auth_seq_id &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ligand_donor_label_seq_id &lt;chr&gt;, ligand_donor_id &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ligand_label_asym_id &lt;chr&gt;, ligand_atom_id &lt;chr&gt;, ligand_id &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ligand_entity_id &lt;chr&gt;, ligand_bond_is_covalent_or_coordinating &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   secondary_structure &lt;chr&gt;, unmodeled_structure &lt;chr&gt;, …</span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan-Philipp Quast, Dina Schuster, ETH Zurich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

