<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ protti</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">protti</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_analysis_dose_response_workflow.html">Dose-Response Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/data_analysis_single_dose_treatment_workflow.html">Single Dose Treatment Data Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/input_preparation_workflow.html">Input Preparation Workflow</a>
    </li>
    <li>
      <a href="../articles/protein_structure_workflow.html">Protein Structure Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/quality_control_workflow.html">Qualtiy Control (QC) Workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jpquast/protti/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="analyse_functional_network.html">analyse_functional_network()</a></code> </p>
        </td>
        <td><p>Analyse protein interaction network for significant hits</p></td>
      </tr><tr><td>
          <p><code><a href="anova_protti.html">anova_protti()</a></code> </p>
        </td>
        <td><p>Perform ANOVA</p></td>
      </tr><tr><td>
          <p><code><a href="assign_missingness.html">assign_missingness()</a></code> </p>
        </td>
        <td><p>Assignment of missingness types</p></td>
      </tr><tr><td>
          <p><code><a href="assign_peptide_type.html">assign_peptide_type()</a></code> </p>
        </td>
        <td><p>Assign peptide type</p></td>
      </tr><tr><td>
          <p><code><a href="barcode_plot.html">barcode_plot()</a></code> </p>
        </td>
        <td><p>Barcode plot</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_aa_scores.html">calculate_aa_scores()</a></code> </p>
        </td>
        <td><p>Calculate scores for each amino acid position in a protein sequence</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_diff_abundance.html">calculate_diff_abundance()</a></code> </p>
        </td>
        <td><p>Calculate differential abundance between conditions</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_go_enrichment.html">calculate_go_enrichment()</a></code> </p>
        </td>
        <td><p>Perform gene ontology enrichment analysis</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_imputation.html">calculate_imputation()</a></code> </p>
        </td>
        <td><p>Sampling of values for imputation</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_kegg_enrichment.html">calculate_kegg_enrichment()</a></code> </p>
        </td>
        <td><p>Perform KEGG pathway enrichment analysis</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_protein_abundance.html">calculate_protein_abundance()</a></code> </p>
        </td>
        <td><p>Label-free protein quantification</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_sequence_coverage.html">calculate_sequence_coverage()</a></code> </p>
        </td>
        <td><p>Protein sequence coverage</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_treatment_enrichment.html">calculate_treatment_enrichment()</a></code> </p>
        </td>
        <td><p>Check treatment enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="create_queue.html">create_queue()</a></code> </p>
        </td>
        <td><p>Creates a mass spectrometer queue for Xcalibur</p></td>
      </tr><tr><td>
          <p><code><a href="create_structure_contact_map.html">create_structure_contact_map()</a></code> </p>
        </td>
        <td><p>Creates a contact map of all atoms from a structure file</p></td>
      </tr><tr><td>
          <p><code><a href="create_synthetic_data.html">create_synthetic_data()</a></code> </p>
        </td>
        <td><p>Creates a synthetic limited proteolysis proteomics dataset</p></td>
      </tr><tr><td>
          <p><code><a href="drc_4p.html">drc_4p()</a></code> </p>
        </td>
        <td><p>Dose response curve helper function</p></td>
      </tr><tr><td>
          <p><code><a href="drc_4p_plot.html">drc_4p_plot()</a></code> </p>
        </td>
        <td><p>Plotting of four-parameter dose response curves</p></td>
      </tr><tr><td>
          <p><code><a href="extract_metal_binders.html">extract_metal_binders()</a></code> </p>
        </td>
        <td><p>Extract metal-binding protein information from UniProt</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_alphafold_aligned_error.html">fetch_alphafold_aligned_error()</a></code> </p>
        </td>
        <td><p>Fetch AlphaFold aligned error</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_alphafold_prediction.html">fetch_alphafold_prediction()</a></code> </p>
        </td>
        <td><p>Fetch AlphaFold prediction</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_chebi.html">fetch_chebi()</a></code> </p>
        </td>
        <td><p>Fetch ChEBI database information</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_eco.html">fetch_eco()</a></code> </p>
        </td>
        <td><p>Fetch evidence &amp; conclusion ontology</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_go.html">fetch_go()</a></code> </p>
        </td>
        <td><p>Fetch gene ontology information from geneontology.org</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_kegg.html">fetch_kegg()</a></code> </p>
        </td>
        <td><p>Fetch KEGG pathway data from KEGG</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_metal_pdb.html">fetch_metal_pdb()</a></code> </p>
        </td>
        <td><p>Fetch structural information about protein-metal binding from MetalPDB</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_mobidb.html">fetch_mobidb()</a></code> </p>
        </td>
        <td><p>Fetch protein disorder and mobility information from MobiDB</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_pdb.html">fetch_pdb()</a></code> </p>
        </td>
        <td><p>Fetch structure information from RCSB</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_pdb_structure.html">fetch_pdb_structure()</a></code> </p>
        </td>
        <td><p>Fetch PDB structure atom data from RCSB</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_quickgo.html">fetch_quickgo()</a></code> </p>
        </td>
        <td><p>Fetch information from the QuickGO API</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_uniprot.html">fetch_uniprot()</a></code> </p>
        </td>
        <td><p>Fetch protein data from UniProt</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_uniprot_proteome.html">fetch_uniprot_proteome()</a></code> </p>
        </td>
        <td><p>Fetch proteome data from UniProt</p></td>
      </tr><tr><td>
          <p><code><a href="filter_cv.html">filter_cv()</a></code> </p>
        </td>
        <td><p>Data filtering based on coefficients of variation (CV)</p></td>
      </tr><tr><td>
          <p><code><a href="find_all_subs.html">find_all_subs()</a></code> </p>
        </td>
        <td><p>Find all sub IDs of an ID in a network</p></td>
      </tr><tr><td>
          <p><code><a href="find_chebis.html">find_chebis()</a></code> </p>
        </td>
        <td><p>Find ChEBI IDs for name patterns</p></td>
      </tr><tr><td>
          <p><code><a href="find_peptide.html">find_peptide()</a></code> </p>
        </td>
        <td><p>Find peptide location</p></td>
      </tr><tr><td>
          <p><code><a href="find_peptide_in_structure.html">find_peptide_in_structure()</a></code> </p>
        </td>
        <td><p>Finds peptide positions in a PDB structure based on positional matching</p></td>
      </tr><tr><td>
          <p><code><a href="fit_drc_4p.html">fit_drc_4p()</a></code> </p>
        </td>
        <td><p>Fitting four-parameter dose response curves</p></td>
      </tr><tr><td>
          <p><code><a href="impute.html">impute()</a></code> </p>
        </td>
        <td><p>Imputation of missing values</p></td>
      </tr><tr><td>
          <p><code><a href="map_peptides_on_structure.html">map_peptides_on_structure()</a></code> </p>
        </td>
        <td><p>Maps peptides onto a PDB structure or AlphaFold prediction</p></td>
      </tr><tr><td>
          <p><code><a href="metal_chebi_uniprot.html">metal_chebi_uniprot</a></code> </p>
        </td>
        <td><p>List of metal-related ChEBI IDs in UniProt</p></td>
      </tr><tr><td>
          <p><code><a href="metal_go_slim_subset.html">metal_go_slim_subset</a></code> </p>
        </td>
        <td><p>Molecular function gene ontology metal subset</p></td>
      </tr><tr><td>
          <p><code><a href="metal_list.html">metal_list</a></code> </p>
        </td>
        <td><p>List of metals</p></td>
      </tr><tr><td>
          <p><code><a href="normalise.html">normalise()</a></code> </p>
        </td>
        <td><p>Intensity normalisation</p></td>
      </tr><tr><td>
          <p><code><a href="parallel_create_structure_contact_map.html">parallel_create_structure_contact_map()</a></code> </p>
        </td>
        <td><p>Creates a contact map of all atoms from a structure file (using parallel processing)</p></td>
      </tr><tr><td>
          <p><code><a href="parallel_fit_drc_4p.html">parallel_fit_drc_4p()</a></code> </p>
        </td>
        <td><p>Fitting four-parameter dose response curves (using parallel processing)</p></td>
      </tr><tr><td>
          <p><code><a href="peptide_profile_plot.html">peptide_profile_plot()</a></code> </p>
        </td>
        <td><p>Peptide abundance profile plot</p></td>
      </tr><tr><td>
          <p><code><a href="predict_alphafold_domain.html">predict_alphafold_domain()</a></code> </p>
        </td>
        <td><p>Predict protein domains of AlphaFold predictions</p></td>
      </tr><tr><td>
          <p><code><a href="protti_colours.html">protti_colours</a></code> </p>
        </td>
        <td><p>Colour scheme for protti</p></td>
      </tr><tr><td>
          <p><code><a href="ptsi_pgk.html">ptsi_pgk</a></code> </p>
        </td>
        <td><p>Structural analysis example data</p></td>
      </tr><tr><td>
          <p><code><a href="pval_distribution_plot.html">pval_distribution_plot()</a></code> </p>
        </td>
        <td><p>Plot histogram of p-value distribution</p></td>
      </tr><tr><td>
          <p><code><a href="qc_charge_states.html">qc_charge_states()</a></code> </p>
        </td>
        <td><p>Check charge state distribution</p></td>
      </tr><tr><td>
          <p><code><a href="qc_contaminants.html">qc_contaminants()</a></code> </p>
        </td>
        <td><p>Percentage of contaminants per sample</p></td>
      </tr><tr><td>
          <p><code><a href="qc_cvs.html">qc_cvs()</a></code> </p>
        </td>
        <td><p>Check CV distribution</p></td>
      </tr><tr><td>
          <p><code><a href="qc_data_completeness.html">qc_data_completeness()</a></code> </p>
        </td>
        <td><p>Data completeness</p></td>
      </tr><tr><td>
          <p><code><a href="qc_ids.html">qc_ids()</a></code> </p>
        </td>
        <td><p>Check number of precursor, peptide or protein IDs</p></td>
      </tr><tr><td>
          <p><code><a href="qc_intensity_distribution.html">qc_intensity_distribution()</a></code> </p>
        </td>
        <td><p>Check intensity distribution per sample and overall</p></td>
      </tr><tr><td>
          <p><code><a href="qc_median_intensities.html">qc_median_intensities()</a></code> </p>
        </td>
        <td><p>Median run intensities</p></td>
      </tr><tr><td>
          <p><code><a href="qc_missed_cleavages.html">qc_missed_cleavages()</a></code> </p>
        </td>
        <td><p>Check missed cleavages</p></td>
      </tr><tr><td>
          <p><code><a href="qc_pca.html">qc_pca()</a></code> </p>
        </td>
        <td><p>Plot principal component analysis</p></td>
      </tr><tr><td>
          <p><code><a href="qc_peak_width.html">qc_peak_width()</a></code> </p>
        </td>
        <td><p>Peak width over retention time</p></td>
      </tr><tr><td>
          <p><code><a href="qc_peptide_type.html">qc_peptide_type()</a></code> </p>
        </td>
        <td><p>Check peptide type percentage share</p></td>
      </tr><tr><td>
          <p><code><a href="qc_proteome_coverage.html">qc_proteome_coverage()</a></code> </p>
        </td>
        <td><p>Proteome coverage per sample and total</p></td>
      </tr><tr><td>
          <p><code><a href="qc_ranked_intensities.html">qc_ranked_intensities()</a></code> </p>
        </td>
        <td><p>Check ranked intensities</p></td>
      </tr><tr><td>
          <p><code><a href="qc_sample_correlation.html">qc_sample_correlation()</a></code> </p>
        </td>
        <td><p>Correlation based hirachical clustering of samples</p></td>
      </tr><tr><td>
          <p><code><a href="qc_sequence_coverage.html">qc_sequence_coverage()</a></code> </p>
        </td>
        <td><p>Protein coverage distribution</p></td>
      </tr><tr><td>
          <p><code><a href="randomise_queue.html">randomise_queue()</a></code> </p>
        </td>
        <td><p>Randomise samples in MS queue</p></td>
      </tr><tr><td>
          <p><code><a href="rapamycin_10uM.html">rapamycin_10uM</a></code> </p>
        </td>
        <td><p>Rapamycin 10 uM example data</p></td>
      </tr><tr><td>
          <p><code><a href="rapamycin_dose_response.html">rapamycin_dose_response</a></code> </p>
        </td>
        <td><p>Rapamycin dose response example data</p></td>
      </tr><tr><td>
          <p><code><a href="read_protti.html">read_protti()</a></code> </p>
        </td>
        <td><p>Read, clean and convert</p></td>
      </tr><tr><td>
          <p><code><a href="replace_identified_by_x.html">replace_identified_by_x()</a></code> </p>
        </td>
        <td><p>Replace identified positions in protein sequence by "x"</p></td>
      </tr><tr><td>
          <p><code><a href="scale_protti.html">scale_protti()</a></code> </p>
        </td>
        <td><p>Scaling a vector</p></td>
      </tr><tr><td>
          <p><code><a href="split_metal_name.html">split_metal_name()</a></code> </p>
        </td>
        <td><p>Convert metal names to search pattern</p></td>
      </tr><tr><td>
          <p><code><a href="try_query.html">try_query()</a></code> </p>
        </td>
        <td><p>Query from URL</p></td>
      </tr><tr><td>
          <p><code><a href="ttest_protti.html">ttest_protti()</a></code> </p>
        </td>
        <td><p>Perform Welch's t-test</p></td>
      </tr><tr><td>
          <p><code><a href="viridis_colours.html">viridis_colours</a></code> </p>
        </td>
        <td><p>Viridis colour scheme</p></td>
      </tr><tr><td>
          <p><code><a href="volcano_plot.html">volcano_plot()</a></code> </p>
        </td>
        <td><p>Volcano plot</p></td>
      </tr><tr><td>
          <p><code><a href="woods_plot.html">woods_plot()</a></code> </p>
        </td>
        <td><p>Woods' plot</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan-Philipp Quast, Dina Schuster, ETH Zurich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

